FaunaDB.Client/Errors/FaunaException.cs

```diff
@@ -10,114 +8,158 @@ namespace FaunaDB.Errors
-            queryErrorResponse.Errors;
+        public string Code { get; private set; }
 
-        public int StatusCode =>
-            queryErrorResponse.StatusCode;
+        /// <summary>
+        /// Array of errors sent by the server.
+        /// </summary>
+        public string[] Position { get; private set; }
 
-        protected FaunaException(QueryErrorResponse response) : base(CreateMessage(response.Errors))
+        public FaunaException(int httpStatusCode, string code, string message, string[] position) : base(message)
         {
-            queryErrorResponse = response;
+            Code = code;
+            HttpStatusCode = httpStatusCode;
+            Position = position;
         }
+    }
 
-        protected FaunaException(string message) : base(message) { }
+    public class InvalidExpression : FaunaException
+    {
+        public InvalidExpression(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvalidExpression, message, position) { }
+    }
 
-        private static string CreateMessage(IReadOnlyList<QueryError> errors) =>
-            string.Join(", ", from error in errors select $"{error.Code}: {error.Description}");
-   }
+    public class InvalidRef : FaunaException
+    {
+        public InvalidRef(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvaliRef, message, position) { }
+    }
 
-    /// <summary>
-    /// HTTP 400 error.
-    /// An exception thrown if FaunaDB cannot evaluate a query.
-    /// <para>HTTP 400 error.</para>
-    /// </summary>
-    public class BadRequest : FaunaException
+    public class InvalidUrlParameter : FaunaException
     {
-        internal BadRequest(QueryErrorResponse response)
-            : base(response) { }
+        public InvalidUrlParameter(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvaliRef, message, position) { }
     }
 
-    /// <summary>
-    /// HTTP 401 error.
-    /// An exception thrown if FaunaDB responds with an HTTP 401 (Unauthorized).
-    /// <para>HTTP 401 error.</para>
-    /// </summary>
-    public class Unauthorized : FaunaException
+    public class SchemaNotFound : FaunaException
     {
-        internal Unauthorized(QueryErrorResponse response)
-            : base(response) { }
+        public SchemaNotFound(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvaliRef, message, position) { }
+    }
+
+    public class InstanceAlreadyExists : FaunaException
+    {
+        public InstanceAlreadyExists(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvaliRef, message, position) { }
+    }
+
+    public class ValidationFailed : FaunaException
+    {
+        public ValidationFailed(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.ValidationFailed, message, position) { }
```

Comment:
There is an internal `failure` object that needs to be added here:
https://github.com/fauna/faunadb-jvm/blob/OSS-864/faunadb-java/src/main/java/com/faunadb/client/FaunaClient.java#L570